packages/controllers/src/cronjob/CronjobController.ts

```diff
@@ -0,0 +1,342 @@
+   * @param job - Cronjob specification.
+   */
+  private schedule(job: Cronjob) {
+    if (this._timers.has(job.id)) {
+      return;
+    }
+    const parsed = parseCronExpression(job.expression);
+    const next = parsed.next();
+    const now = new Date();
+    const ms = next.getTime() - now.getTime();
+
+    // Don't schedule this job yet as it is too far in the future
+    if (ms > DAILY_TIMEOUT) {
+      return;
+    }
+
+    const timer = new Timer(ms);
+    timer.start(() => {
+      this.executeCronjob(job);
+      this._timers.delete(job.id);
+      this.schedule(job);
+    });
+
+    this.updateJobLastRunState(job.id, 0); // 0 for init, never ran actually
+    this._timers.set(job.id, timer);
+    this._snapIds.set(job.id, job.snapId);
+  }
+
+  /**
+   * Execute job.
+   *
+   * @param job - Cronjob specification.
+   */
+  private executeCronjob(job: Cronjob) {
+    this.updateJobLastRunState(job.id, Date.now());
+    this._messenger.call('SnapController:handleRequest', {
+      snapId: job.snapId,
+      origin: '',
+      handler: HandlerType.OnCronjob,
+      request: job.request,
+    });
+  }
+
+  /**
+   * Unregister all jobs related to the given snapId.
+   *
+   * @param snapId - ID of a snap.
+   */
+  unregister(snapId: SnapId) {
+    const jobs = [...this._snapIds.entries()].filter(
+      ([_, jobSnapId]) => jobSnapId === snapId,
+    );
+
+    if (jobs.length) {
+      jobs.forEach(([id]) => {
+        const timer = this._timers.get(id);
+        if (timer) {
+          timer.cancel();
+          this._timers.delete(id);
+          this._snapIds.delete(id);
+        }
+      });
+    }
+  }
+
+  /**
+   * Update time of a last run for the Cronjob specified by ID.
+   *
+   * @param jobId - ID of a cron job.
+   * @param lastRun - Unix timestamp when the job was last ran.
+   */
+  private updateJobLastRunState(jobId: string, lastRun: number) {
+    this.update((state) => {
+      state.jobs[jobId] = {
+        lastRun,
+      };
+    });
+  }
+
+  /**
+   * Runs every 24 hours to check if new jobs need to be scheduled.
+   *
+   * This is necesary for longer running jobs that execute with more than 24 hours between them.
+   */
+  async dailyCheckin() {
+    const jobs = await this.getAllJobs();
+    jobs.forEach((job) => {
+      const parsed = parseCronExpression(job.expression);
+      const lastRun = this.state.jobs[job.id]?.lastRun;
+      // If a job was supposed to run while we were shut down but wasn't we run it now
+      if (
+        lastRun !== undefined &&
+        parsed.hasPrev() &&
+        parsed.prev().getTime() > lastRun
+      ) {
+        this.executeCronjob(job);
+      }
+
+      // Try scheduling, will fail if an existing scheduled job is found
+      this.schedule(job);
+    });
+    this._dailyTimer = new Timer(DAILY_TIMEOUT);
+    this._dailyTimer.start(() => this.dailyCheckin());
+  }
+
+  /**
+   * Run controller teardown process and unsubscribe from Snap events.
+   */
+  destroy() {
```

Comment:
Should we cancel any running jobs here?

Reply:
This might actually be a good idea to have that implemented. Since we have `unregister` method I made use of it and unregistered everything for each snap. I also added test to cover it.

I'm not fully aware where and when destroy is used across the project so @FrederikBolding let me know if you have any concerns about this.

Reply:
Nope, that sounds like a good idea.